### ISSUE TITLE
Inconsistent results based on number of threads used

### ISSUE THREAD

**Issue by snapbug** [pytorch/pytorch#2201]
Our lab has been having a minor reproducibility crisis - training the same model on different computers was leading to different results, and the differences were substantial. After trying a large amount of things, long story short, setting `OMP_NUM_THREADS` and `MKL_NUM_THREADS` changes the results of the network, even when controlling random seeds etc.

The effect is observable in the example networks from pytorch/example. To test, I spun up a t2.medium instance on ec2 running amazon linux (to be entirely sure that it wasn't some other environment issue), and installed miniconda. After which I installed pytorch and torch vision:

```
[ec2-user@ip-172-31-38-212 ~]$ conda install pytorch torchvision -c soumith
Fetching package metadata ...........
Solving package specifications: .
Warning: 2 possible package resolutions (only showing differing packages):
  - soumith::pytorch-0.1.12-py360_2cu75
  - soumith::pytorch-0.1.12-py36_2cu75

Package plan for installation in environment /home/ec2-user/miniconda3:

The following NEW packages will be INSTALLED:

    freetype:    2.5.5-2
    jbig:        2.1-0
    jpeg:        9b-0
    libgcc:      5.2.0-0
    libpng:      1.6.27-0
    libtiff:     4.0.6-3
    mkl:         2017.0.3-0
    numpy:       1.13.1-py36_0
    olefile:     0.44-py36_0
    pillow:      4.2.1-py36_0
    pytorch:     0.1.12-py360_2cu75 soumith
    torchvision: 0.1.8-py36_2       soumith

The following packages will be UPDATED:

    conda:       4.3.14-py36_0              --> 4.3.22-py36_0
```

I then cloned the pytorch/examples repository:

```
[ec2-user@ip-172-31-38-212 ~]$ cd examples
[ec2-user@ip-172-31-38-212 examples]$ git rev-parse HEAD
0722b2f4f73a220109dd2c0681d37df25f4a4116
```

From there I ran the mnist example:

```
[ec2-user@ip-172-31-38-212 examples]$ cd mnist
[ec2-user@ip-172-31-38-212 mnist]$ python main.py --epochs 1 --no-cuda
# output omitted
Train Epoch: 1 [58880/60000 (98%)]	Loss: 0.338396
Train Epoch: 1 [59520/60000 (99%)]	Loss: 0.666101

Test set: Average loss: 0.1978, Accuracy: 9436/10000 (94%)
```

I then re-ran the example, having set `OMP_NUM_THREADS` and `MKL_NUM_THREADS`:

```
[ec2-user@ip-172-31-38-212 mnist]$ export OMP_NUM_THREADS=1
[ec2-user@ip-172-31-38-212 mnist]$ export MKL_NUM_THREADS=1
[ec2-user@ip-172-31-38-212 mnist]$ python main.py --epochs 1 --no-cuda
# output omitted
Train Epoch: 1 [58880/60000 (98%)]	Loss: 0.329742
Train Epoch: 1 [59520/60000 (99%)]	Loss: 0.654222

Test set: Average loss: 0.2003, Accuracy: 9438/10000 (94%)
```

And the loss/accuracy are different after training.

In this example the differences are small, but in the network we're running the differences have been quite large, covering a range of the last few years of progress on our problem. And really it shouldn't matter how many threads are being run, the resulting model should be the same for the same network with the same seeds.

**Comment by albanD** [comment]
Hi,
I have already seen this kind of behavior for cuda operations that are non deterministic.
I guess changing the number of threads here can also change the order of the floating point operations, leading to different values. Following these by gradient descent then leads to completely different models.
Can you check what is the order of magnitude of the difference for a single operation? If it is at the level of float precision then this might be it, if they are exactly the same (to the last bit) then there might be something else wrong.

**Comment by snapbug** [comment]
I'd be happy to help, but I'm not sure how would I go about doing that?

In the time between comments I've run both conditions again with `--batch-size 1 --test-batch-size 1 --log-interval 1`, with no differences (I suspect unsurprisingly) in output between the thread settings.

**Comment by snapbug** [comment]
I'm closing this issue, as the results do seem to be based on ordering of floating point operations, as in the results from the forward passes started differing after a few epochs by inspecting some `float.hex` values, indeed suggesting minor differences from the ordering of ops.

It was initially surprising that the number of threads made a difference (my usual field doesn't exhibit anywhere near this level of sensitivity), as does the architecture being run on (which was a further issue we encountered).
